forge/test/operators/utils/plan.py

```diff
@@ -0,0 +1,465 @@
+        if self.operators is not None:
+            self.operators = PytestParamsUtils.strip_param_sets(self.operators)
+        if self.input_sources is not None:
+            self.input_sources = PytestParamsUtils.strip_param_sets(self.input_sources)
+        if self.input_shapes is not None:
+            self.input_shapes = PytestParamsUtils.strip_param_sets(self.input_shapes)
+        if self.dev_data_formats is not None:
+            self.dev_data_formats = PytestParamsUtils.strip_param_sets(self.dev_data_formats)
+        if self.math_fidelities is not None:
+            self.math_fidelities = PytestParamsUtils.strip_param_sets(self.math_fidelities)
+        if self.kwargs is not None and not isinstance(self.kwargs, types.FunctionType):
+            self.kwargs = PytestParamsUtils.strip_param_sets(self.kwargs)
+
+
+@dataclass
+class TestPlan:
+    '''
+    Define test plan for the operator testing. Define failing rules for the tests.
+    
+    Args:
+        collections: List of test collections
+        failing_rules: List of failing rules
+    '''
+
+    __test__ = False  # Avoid collecting TestPlan as a pytest test
+
+    collections: Optional[List[TestCollection]] = None
+    failing_rules: Optional[List[TestCollection]] = None
+
+    def _check_test_failing(
+        self,
+        test_vector: TestVector,
+    ) -> Optional[TestResultFailing]:
+        '''Check if the test is failing based on the test plan
+        
+        Args:
+            test_vector: Test vector with all the parameters
+        '''
+
+        failing_result = None
+
+        for failing_rule in self.failing_rules:
+            if TestPlanUtils.test_vector_in_collection(test_vector, failing_rule):
+                if failing_rule.failing_reason is not None or failing_rule.skip_reason is not None:
+                    failing_result = TestResultFailing(failing_rule.failing_reason, failing_rule.skip_reason)
+                else:
+                    # logger.debug(f"Test should pass: {test_vector.get_id()}")
+                    failing_result = None
+
+        return failing_result
+
+    def generate(self) -> Generator[TestVector, None, None]:
+        '''Generate test vectors based on the test plan'''
+
+        for test_collection in self.collections:
+
+            dev_data_formats = test_collection.dev_data_formats
+            if dev_data_formats is None:
+                dev_data_formats = [None]
+
+            math_fidelities = test_collection.math_fidelities
+            if math_fidelities is None:
+                math_fidelities = [None]
+
+            kwargs_list = test_collection.kwargs
+            if kwargs_list is None:
+                kwargs_list = [None]
+
+            for input_operator in test_collection.operators:
+                for input_source in test_collection.input_sources:
+                    for input_shape in test_collection.input_shapes:
+                        if isinstance(input_shape, ParameterSet):
```

Comment:
Yes, you are right, it was unnecessary repetitive. Actually it was completely redundant since ParameterSet has been already removed in TestCollection.__post_init__, so I just removed it from loops.

For the context, the idea of ParameterSet here is to allow defining collections in two ways:
  - list of values
  - list of ParameterSet (values + pytest marks)
In the future this will probably be restricted to just first variant.